addon/-private/system/many-array.js

```diff
@@ -141,31 +148,64 @@ export default Ember.Object.extend(Ember.MutableArray, Ember.Evented, {
     return this.currentState[index].getRecord();
   },
 
-  flushCanonical(isInitialized = true) {
-    let toSet = this.canonicalState;
+  flushCanonical() {
+    // It’s possible the parent side of the relationship may have been unloaded by this point
+    if (!_objectIsAlive(this)) {
+      return;
+    }
+    //TODO make this smarter, currently its plenty stupid
+    let toSet = this.canonicalState.filter((internalModel) => !internalModel.isDeleted());
 
     //a hack for not removing new records
     //TODO remove once we have proper diffing
-    let newRecords = this.currentState.filter(
+    const newRecords = this.currentState.filter(
       // only add new records which are not yet in the canonical state of this
       // relationship (a new record can be in the canonical state if it has
       // been 'acknowleged' to be in the relationship via a store.push)
       (internalModel) => internalModel.isNew() && toSet.indexOf(internalModel) === -1
     );
     toSet = toSet.concat(newRecords);
-    let oldLength = this.length;
-    this.arrayContentWillChange(0, this.length, toSet.length);
-    // It’s possible the parent side of the relationship may have been unloaded by this point
-    if (_objectIsAlive(this)) {
-      this.set('length', toSet.length);
+    const oldLength = this.length;
+    const newLength = toSet.length;
+
+    const shortestLength = Math.min(oldLength, newLength);
+
+    let firstChangeIndex = -1; // -1 signifies no changes
+    // find the first change
+    const currentArray = this.currentState;
+    for (let i=0; i<shortestLength; i++) {
+      // compare each item in the array
+      if (currentArray[i] !== toSet[i]) {
+        firstChangeIndex = i;
+        break;
+      }
     }
-    this.currentState = toSet;
-    this.arrayContentDidChange(0, oldLength, this.length);
-
-    if (isInitialized) {
-      //TODO Figure out to notify only on additions and maybe only if unloaded
+    if (firstChangeIndex === -1) {
+      // no change found in the matching part of the arrays
+      if (newLength !== oldLength) {
+        firstChangeIndex = shortestLength;
+      }
+    }
+    if (firstChangeIndex !== -1) {
+      // we found a change, find the end of the change
+      let unchangedEndBlockLength = 0;
+      // walk back from the end of both arrays until we find a change
+      for (let i=1; i<shortestLength; i++) {
+        // compare each item in the array
+        if (currentArray[oldLength-i] !== toSet[newLength-i]) {
+          unchangedEndBlockLength = i-1;
+          break;
+        }
+      }
+      const added = newLength - unchangedEndBlockLength - firstChangeIndex;
+      const removed = oldLength - unchangedEndBlockLength - firstChangeIndex;
+      this.arrayContentWillChange(firstChangeIndex, removed, added);
```

Comment:
if nothing changed, should be bail out here entirely? Seems like we don't want to emit any events in that case.